File: solutions/0213-house-robber-ii/solution.py
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n = len(nums)
        if n <= 3:
            return max(nums)
        dp1= [-1]*(n-1)
        dp2= [-1]*(n-1)
        num1 = nums[:n-1]
        num2 = nums[1:n]
        dp1[0] = num1[0]
        dp2[0] = num2[0]
   

        def f(i,dp,arr):
            if i < 0:
                return 0
            if i == 0:
                return arr[0]
            if dp[i] != -1:
                return dp[i]
            dp [i] = max((arr[i]+f(i-2,dp,arr)),f(i-1,dp,arr))
            return dp[i]
        
    
        return max(f(n-2,dp1,num1),f(n-2,dp2,num2))

        

File: solutions/0213-house-robber-ii/solution.java
class Solution {

  

    public int f(int i,int[] dp,int[] nums) {
        if (i == 0) return nums[0];
        if (i == 1) return Math.max(nums[0], nums[1]);

        if (dp[i] != -1) return dp[i];

        dp[i] = Math.max(f(i - 1,dp,nums), f(i - 2,dp,nums)+ nums[i]);
        return dp[i];
    }

    public int rob(int[] nums) {
        int n = nums.length;

        if (n == 0) return 0;
        if (n == 1) return nums[0];
        int[] nums1 = Arrays.copyOfRange(nums, 0, n-1);
        int[] nums2 = Arrays.copyOfRange(nums, 1, n);
        int[] dp1 = new int[n-1];
        int[] dp2 = new int[n-1];
        Arrays.fill(dp1, -1);
        Arrays.fill(dp2, -1);

        return Math.max(f(n - 2,dp1,nums1) , f(n-2,dp2,nums2));
    }
}
